### PR TITLE
8279640: ListView with null SelectionModel/FocusModel throws NPE

### PR DESCRIPTION
**Pull request by Maran23**
This PR fixes a bunch of NPEs when a null `SelectionModel` or `FocusModel` is set on a `ListView`.

The following NPEs are fixed (all are also covered by exactly one test case):
NPEs with null selection model:
- Mouse click on a `ListCell`
- SPACE key press
- KP_UP (arrow up) key press
- HOME key press
- END key press
- BACK_SLASH + CTRL key press

NPEs with null focus model:
- SPACE key press
- Select an items: getSelectionModel().select(1)
- Clear-Select an item and add one after: `listView.getSelectionModel().clearAndSelect(1); listView.getItems().add("3");`

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8279640](https://bugs.openjdk.org/browse/JDK-8279640): ListView with null SelectionModel/FocusModel throws NPE

### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - Author) 🔄 Re-review required (review applies to [9146e080](https://git.openjdk.org/jfx/pull/711/files/9146e08022316d71ca6584260d9a1273940ab379))
 * [Jeanette Winzenburg](https://openjdk.org/census#fastegal) (@kleopatra - **Reviewer**)
 * [Ajit Ghaisas](https://openjdk.org/census#aghaisas) (@aghaisas - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx pull/711/head:pull/711` \
`$ git checkout pull/711`

Update a local copy of the PR: \
`$ git checkout pull/711` \
`$ git pull https://git.openjdk.org/jfx pull/711/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 711`

View PR using the GUI difftool: \
`$ git pr show -t 711`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/711.diff">https://git.openjdk.org/jfx/pull/711.diff</a>

</details>

### CODE REVIEW

**Comment by hjohn** [comment]
Was it considered to not allow these models to be set to `null` instead? Or to use a Null Object? I'd much prefer this instead of polluting the code with `null` checks on every access of these models. 

Furthermore, the API makes no mention what is supposed to happen when these models are set to `null`, so this seems undefined at this time.  If `null` is supposed to be valid then I think the documentation should clearly state what this means.  Having no focus model for example should probably mean that nothing can ever get focus and that the focused item is always null and index is always -1.  Instead of adding null checks to achieve this however I'd use a `NullFocusModel` instance as it is  easy to forget a `null` check (also when making future changes).

Same goes for the selection model; if set to `null` it should not allow any kind of selection, which again seems best achieved by installing a model that ignores all such actions and always runs empty values.

**Comment by Maran23** [comment]
> Same goes for the selection model; if set to `null` it should not allow any kind of selection, which again seems best achieved by installing a model that ignores all such actions and always returns empty values.

We had a similar discussion on https://github.com/openjdk/jfx/pull/557. We came to the conclusion that we allow null as a selection or focusmodel and don't set a noop selection model or something of this kind.

The biggest problem with this is the following example:
I as a developer set the selection model to null via `setSelectionModel(null)`. Now if the code silently set the selection model to an empty noop selection model, we won't get null when calling `getSelectionModel()`, which a developer would expect.

Also from a look of the code, even if we use a noop focus model, we would still the focus because of the `anchor` stuff, which is set in the `ListViewBehaviour`. 
If we do a null check and fast return like now, they won't be set -> there are not visible. So it might not even fix our problem but makes even more.

**Comment by hjohn** [comment]
> We had a similar discussion on #557. We came to the conclusion that we allow null as a selection or focusmodel and don't set a noop selection model or something of this kind.

Even earlier though in https://github.com/javafxports/openjdk-jfx/issues/569 a `NONE` model was suggested.  And I still think that makes much more sense, because in effect the "behavior" of the `NONE` model is now spread out over dozens of places in the code (as the `null` checks are was is defining the behavior) instead of in a single place.

> The biggest problem with this is the following example: I as a developer set the selection model to null via `setSelectionModel(null)`. Now if the code silently set the selection model to an empty noop selection model, we won't get null when calling `getSelectionModel()`, which a developer would expect.

This is an example from `ListView` which already does something custom because the property is lazily instantiated:

    public final MultipleSelectionModel<T> getSelectionModel() {
        return selectionModel == null ? null : selectionModel.get();
    }

Just write that as:

    public final MultipleSelectionModel<T> getSelectionModel() {
        return selectionModel == null ? null : 
             selectionModel.get() == NONE_SELECTION_MODEL ? null : selectionModel.get();
    }
    
`NONE_SELECTION_MODEL` is stateless so only a single instance is needed.
 
> Also from a look of the code, even if we use a noop focus model, we would still the focus because of the `anchor` stuff, which is set in the `ListViewBehaviour`. If we do a null check and fast return like now, they won't be set -> there are not visible. So it might not even fix our problem but makes even more.

A check can also be done to see if something matches the `NONE` model, just like you can check for `null`, so you can still fast return in special cases.

The chosen approach works as well of course.

**Comment by Maran23** [comment]
> ```
> public final MultipleSelectionModel<T> getSelectionModel() {
>     return selectionModel == null ? null : 
>          selectionModel.get() == NONE_SELECTION_MODEL ? null : selectionModel.get();
> }
> ```

That would work altough I think it's a bit hacky as when I listen on the `selectionModelProperty` I still would get the noop selection model.
I'm in general not a fan of 'self-changing' properties. When a null value is possible, guards against are needed. 
Unfortunately there is no built-in way to forbid null in Java as whole.

> A check can also be done to see if something matches the `NONE` model, just like you can check for `null`, so you can still fast return in special cases.

But then I don't see any advantage as it makes no difference if we check for NONE model or null. Checks are needed in any case.

**Comment by mstr2** [comment]
> When a null value is possible, guards against are needed. Unfortunately there is no built-in way to forbid null in Java as whole.

I think null checks should only be done to verify the preconditions of an operation, and not to guard against bugs. This means that, when `null` is not a _valid_ value for a property (regardless of whether it is a _possible_ value), no attempt should be made to guard against this in downstream code. In fact, I would argue that doing so is a strong code smell.

The question is whether or not `null` should be a valid value for the `selectionModel` and `focusModel` properties. I think there are good reasons to think that it shouldn't. The primitive property classes (`IntegerProperty`, `BooleanProperty`, etc.) have the same issue, and they deal with this situation by coercing the invalid `null` value to their respective default value.

It's not clear to me whether the `ListCellBehavior.anchor` attached property setter would need any special-casing. When `null` (or `-1` in case of `ListViewBehavior.setAnchor`) is passed in, it seems to remove the anchor. Isn't that what we'd expect?

**Comment by Maran23** [comment]
> The question is whether or not `null` should be a valid value for the `selectionModel` and `focusModel` properties. I think there are good reasons to think that it shouldn't. The primitive property classes (`IntegerProperty`, `BooleanProperty`, etc.) have the same issue, and they deal with this situation by coercing the invalid `null` value to their respective default value.

I think the property classes for primitives are another use case, since those values really can't be null.

I get your point but I think a null `selectionModel` or `focusModel` should be allowed. 
At least an use case would be to have a read-only `ListView` which should not be selectable nor focusable.
If you just set a ListView to uneditable (via `setEditable(..)`), you can still select entries. If you disable the `ListView` it is grey styled which is not desired in this use case. Of course I can change the style, but a null selection and-focusModel may make sense as well.

But I agree to a certain point, maybe it make's sense to fail for some properties when null is set/or use default values (but without changing the property?). Right now some properties in e.g. `Node` can be set to null and it will throw a NPE somewhere (e.g. `setCacheHint(null)`, `setRotationAxis(null)`, `Labeled.setFont(null)` ...). 
But on the other hand a lot of code handle properties with null with some kind of default behaviour (the property itself is never changed). 
Examples for this:
- `Labeled.getTextOverrun()` -> null will be handled as ELLIPSIS in LabeledSkinBase
- `Labeled.getEllipsisString()` -> null will be handled as "" in LabeledSkinBase
- ...

So there is no consistent behaviour for this, but a lot of code already handles null by behaving in some kind of default way without changing the property directly, and I think it might be the best to adapt to this.

**Comment by kevinrushforth** [comment]
I'd like to see this get resolved for JavaFX 20. I note there is a similar issue with `TableView`, which is tracked by [JDK-8187145](https://bugs.openjdk.org/browse/JDK-8187145).

As I mentioned in that JBS issue, I tend to agree that if we were starting today with a blank sheet of paper, we might have disallowed null and defined a "noop" selection model and/or focus model. At this point, though, it would be a breaking change, even though we don't specify the behavior of null, so we ought to make it work in those places where it doesn't.

> So there is no consistent behaviour for this, but a lot of code already handles null by behaving in some kind of default way without changing the property directly, and I think it might be the best to adapt to this.

This seems like the best way forward to me, but let's see what comes out of the review.

As part of this, we should clarify the spec that a null selection (and focus) model is allowed, and say what it does. This might mean a CSR, but I'd want to look at the current docs first.

**Comment by andy-goryachev-oracle** [comment]
Setting selection model to null (i.e. disabling selection) is a fairly frequent operation, @kevinrushforth .
The code must support that.  So no CSR is needed, I think.

**Comment by kevinrushforth** [comment]
I took a look at the bug description for [JDK-8230098](https://bugs.openjdk.org/browse/JDK-8230098), and it should be closed as a duplicate of [JDK-8279640](https://bugs.openjdk.org/browse/JDK-8279640) rather than adding both as solved by this PR (adding additional issues to a PR is used when that PR solves two different, but related issues).

So let's closed JDK-8230098 as a duplicate and you can remove it from this PR with `/issue remove`.

**Comment by kevinrushforth** [comment]
> So no CSR is needed, I think.

Probably you are right if all we end up doing is clarifying existing behavior. For example, if we add a sentence to the effect that "a null selection model disables selection", and that is what already happens today, then a CSR would likely not be needed. Unlike the JDK, which requires a CSR for any non-trivial spec change, we have often added clarifying comments to JavaFX API without a CSR.

**Comment by andy-goryachev-oracle** [comment]
good point.  +1 for clarification; same applies to TableView JDK-8187145.
I wonder if we also need to apply the same treatment to TreeTableView.

**Comment by kevinrushforth** [comment]
Quite possibly. Can you check it as part of fixing [JDK-8187145](https://bugs.openjdk.org/browse/JDK-8187145) and expand that issue to cover TreeTableView if needed?

**Comment by andy-goryachev-oracle** [comment]
> Can you check it as part of fixing [JDK-8187145](https://bugs.openjdk.org/browse/JDK-8187145) and expand that issue to cover TreeTableView if needed?

Absolutely.  Just checked - we have the same problem there.

**Comment by Maran23** [comment]
> As I mentioned in that JBS issue, I tend to agree that if we were starting today with a blank sheet of paper, we might have disallowed null and defined a "noop" selection model and/or focus model. At this point, though, it would be a breaking change, even though we don't specify the behavior of null, so we ought to make it work in those places where it doesn't.

I agree, if we could do it from scratch we can just implement something like a `NonNullObjectProperty` which won't allow null in some way. Or the JDK implements some way of declaring a variable as not nullable. 😃 

And I agree that improving the docs makes sense and more information is always good.

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java [hunk 1]
```diff
@@ -262,6 +262,10 @@ public ListViewBehavior(ListView<T> control) {
     private boolean selectionChanging = false;
 
     private final ListChangeListener<Integer> selectedIndicesListener = c -> {
+        if (getNode().getFocusModel() == null) {
+            return;
+        }
+
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java:268]: line 281 (below): could sm be null there?

**Comment by andy-goryachev-oracle** [comment]
I wonder if a null check should be added to:
- ListViewSkin:381 calls getSelectionModel()

**Comment by andy-goryachev-oracle** [comment]
There is a condition observed while debugging [JDK-8187145](https://bugs.openjdk.org/browse/JDK-8187145),
when a non-null selection model is set, the user creates a non-empty selection, then the selection is set to null.

The UI still shows selected cells - does the same bug exists in ListView?

**Review by andy-goryachev-oracle — CHANGES_REQUESTED** [review]

**Comment by Maran23** [comment]
> ListViewSkin:381 calls getSelectionModel()

probably, same to focusModel here.

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java (hunk 1 above)
Comment (Maran23) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java:268]: no, as this is listener is added to a particular selection model.

**Comment by Maran23** [comment]
> There is a condition observed while debugging [JDK-8187145](https://bugs.openjdk.org/browse/JDK-8187145), when a non-null selection model is set, the user creates a non-empty selection, then the selection is set to null.
> 
> The UI still shows selected cells - does the same bug exists in ListView?

Just tried and it looks fine for `ListView`.

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewMouseInputTest.java [hunk 2]
```diff
@@ -384,4 +385,15 @@ private boolean isAnchor(int index) {
...
+
+    @Test public void testClickWithNullSelectionModelDoesNotThrowNPE() {
+        listView.setSelectionModel(null);
+
+        StageLoader loader = new StageLoader(listView);
+
+        assertDoesNotThrow(()  -> VirtualFlowTestUtils.clickOnRow(listView, 2));
+
+        loader.dispose();
+    }
```
Comment (kleopatra) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewMouseInputTest.java:397]: the stageloader is not disposed if the test fails - so a safer route is to keep it in a field which is then disposed in  @ after (which we decided to do years ago, but tend to forget occasionally :), see f.i. ListViewKeyInputTest

BTW: good to explicitely create the stageloader and __not__ rely on the one created implicitly created by the flowUtils!

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java [hunk 3]
```diff
@@ -2170,6 +2187,31 @@ public void testItemLeak() {
         assertNull("ListView item is not GCed.", itemRef.get());
     }
 
+    @Test
+    public void testSelectWithNullFocusModelDoesNotThrowNPE() {
+        listView.getItems().addAll("1", "2");
+        listView.setFocusModel(null);
+
+        StageLoader stageLoader = new StageLoader(listView);
+
```
Comment (kleopatra) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java:2196]: just to not forget: stageloader again

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java [hunk 4]
```diff
@@ -2170,6 +2187,31 @@ public void testItemLeak() {
...
+        stageLoader.dispose();
+    }
+
+    @Test
+    public void testItemChangeWithNullFocusModelDoesNotThrowNPE() {
+        listView.getItems().addAll("1", "2");
+        listView.setFocusModel(null);
+
+        StageLoader stageLoader = new StageLoader(listView);
+
```
Comment (kleopatra) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/ListViewTest.java:2208]: just to not forget: stageloader again

**Review by kleopatra — CHANGES_REQUESTED** [review]
Fix and tests look rather complete, NPEs are fixed

There is still one unguarded access to the selectionModel in ListViewSkin.queryAccessibleAttributes: don't know if that's ever reached without actually having a selectionModel?

As we have similar NPEs in all our behaviors (of virtualized controls), we might decided here which behavior we want to support and then apply that decision everywhere. In particular what should happen if we have a null selection and a not-null focusModel with

- activate/edit: that happens on the focusedIndex/Cell with selection only an after-thought
- focus navigation (often ctrl-somekey): is inconsistently either supported (ctrl-end) or not (ctrl-pageDown)

Personally, I think that we should support both - we have a separate focusModel so should use it independently of the selectionModel where possible. A good starter for this issue might be to implement the activate such that it doesn't back out on null selection (just leave out the select statement). The consistent support of navigation might be done in a follow-up issue.

**Comment by andy-goryachev-oracle** [comment]
There is also an unguarded access in CellBehaviorBase.selectRows(int,int) : 312 and CellBehaviorBase.sompleSelect(MouseButton,int,boolean) : 274, which are being fixed by #876

I'd suggest to integrate #876 first, followed by this PR, in order to avoid merge conflicts.

**Comment by kleopatra** [comment]
> There is also an unguarded access in CellBehaviorBase.selectRows(int,int) : 312 and CellBehaviorBase.sompleSelect(MouseButton,int,boolean) : 274, which are being fixed by #876
> 
> I'd suggest to integrate #876 first, followed by this PR, in order to avoid merge conflicts.

@andy-goryachev-oracle  hmm .. are those methods reached if selectionModel == null? On skimming across its code, it looks like all callers back out (that is not call it) without selectionModel. Might be wrong, though - hard to tell without documented preconditions, and without tests 😉

**Comment by andy-goryachev-oracle** [comment]
@kleopatra : 
simpleSelect and selectRows are being called from CellBehaviorBase.mousePressed and mouseReleased, so probably yes.
in any way, they are updated in #876 ; fixing them here would likely create a merge conflict.

**Comment by andy-goryachev-oracle** [comment]
Screen Shot 2022-08-22 at 09 37 36

**Comment by kleopatra** [comment]
doSelect backs out if selection or focusModel is null - so we probably don't need to touch the 

> @kleopatra : simpleSelect and selectRows are being called from CellBehaviorBase.mousePressed and mouseReleased, so probably yes. in any way, they are updated in #876 ; fixing them here would likely create a merge conflict.

to be precise: the mouse methods call doSelect which backs out if focus- or selectionModel is null - so doesn't look like we should touch simpleSelect/selectRows. Except maybe by adding a precondition that the selection/focusModel must not be null. But we'll see in review, you might convince me with tests :) 

Here at least cell behavior is not changed, so changes in the other PR wouldn't cause a conflict.

**Comment by andy-goryachev-oracle** [comment]
you are right!  unless, of course, someone writes code to invoke these methods directly.  For example, simpleSelect() is protected, so any Behavior that extends CellBehaviorBase might encounter an NPE.

but again, these are being addressed in another PR.

**Comment by Maran23** [comment]
> you are right! unless, of course, someone writes code to invoke these methods directly. For example, simpleSelect() is protected, so any Behavior that extends CellBehaviorBase might encounter an NPE.

No strong opinion here but I think in this case it doesn't really matter since all behaviours are anyway in `com.sun.javafx` and since `simpleSelect` is called by `doSelect`, there won't be any problem. 
But another null check does not hurt on the other hand.

**Comment by Maran23** [comment]
> Fix and tests look rather complete, NPEs are fixed
> 
> There is still one unguarded access to the selectionModel in ListViewSkin.queryAccessibleAttributes: don't know if that's ever reached without actually having a selectionModel?
> 
> As we have similar NPEs in all our behaviors (of virtualized controls), we might decided here which behavior we want to support and then apply that decision everywhere. In particular what should happen if we have a null selection and a not-null focusModel with
> 
> * activate/edit: that happens on the focusedIndex/Cell with selection only an after-thought
> * focus navigation (often ctrl-somekey): is inconsistently either supported (ctrl-end) or not (ctrl-pageDown)
> 
> Personally, I think that we should support both - we have a separate focusModel so should use it independently of the selectionModel where possible. A good starter for this issue might be to implement the activate such that it doesn't back out on null selection (just leave out the select statement). The consistent support of navigation might be done in a follow-up issue.

`queryAccessibleAttributes` is now fixed, as well as the other things you mentioned. :)

**Review by andy-goryachev-oracle — APPROVED** [review]
lgtm

File: modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java [hunk 5]
```diff
@@ -563,8 +577,11 @@ private void selectPreviousRow() {
             return;
         }
 
+        MultipleSelectionModel<T> selectionModel = getNode().getSelectionModel();
+        if (selectionModel == null) return;
```
Comment (danielpeintner) [inline modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java:581]: VERY MINOR style issue: this seems to be the only place where you use a _one-liner_ return

Comment (Maran23) [reply modules/javafx.controls/src/main/java/com/sun/javafx/scene/control/behavior/ListViewBehavior.java]: Yes, this is unfortunately different everywhere in JavaFX and I usually just adapt to the already existing code, but in this case I didn't. Just changed it.

**Comment by tsayao** [comment]
Just updated to JavaFX 19 and noticed this NPE:

Exception in thread "JavaFX Application Thread" java.lang.NullPointerException: Cannot invoke "javafx.scene.control.TableView$TableViewSelectionModel.getSelectedCells()" because the return value of "javafx.scene.control.TableView.getSelectionModel()" is null

This is when setting SelectionModel to null (because I don't want the TableView to be selectable) and clicking on the headerbar to sort.

**Comment by tsayao** [comment]
> > Just updated to JavaFX 19 (without this fix) and noticed this NPE:
> > Exception in thread "JavaFX Application Thread" java.lang.NullPointerException: Cannot invoke "javafx.scene.control.TableView$TableViewSelectionModel.getSelectedCells()" because the return value of "javafx.scene.control.TableView.getSelectionModel()" is null
> > This is when setting SelectionModel to null (because I don't want the TableView to be selectable) and clicking on the headerbar to sort.
> 
> This PR is only touching the `ListView`, not the `TableView`. The behaviour you mentioned is fixed in another PR: #876

My mistake.

**Comment by kleopatra** [comment]
> 
> Personally, I think that we should support both - we have a separate focusModel so should use it independently of the selectionModel where possible. A good starter for this issue might be to implement the activate such that it doesn't back out on null selection (just leave out the select statement). The consistent support of navigation might be done in a follow-up issue.

Seeing you on a similar path :) Just to summarize, with null selectionModel

- appropriate keys on the focused row starts the edit with null selection model
- focus navigation is left for a future issue 

two suggestions: 

- we might want a test to capture the start edit, f.i. on SPACE
- file an issue covering the focus navigation to not forget?

**Review by kleopatra — APPROVED** [review]
left a comment - which is only tangentially related to the issue to fix here. 

So this looks good. Cool tests for the code in the accessibility block, my learn item of the day :))

**Comment by Maran23** [comment]
> Seeing you on a similar path :) Just to summarize, with null selectionModel
> 
> * appropriate keys on the focused row starts the edit with null selection model
> * focus navigation is left for a future issue
> 
> two suggestions:
> 
> * we might want a test to capture the start edit, f.i. on SPACE
> * file an issue covering the focus navigation to not forget?

I agree, we should look into this. 
The question is: Can we activate/edit a focused cell without a `selectionModel` (only `focusModel` set)?
And fix all inconsistencies in e.g. navigation.

I filed https://bugs.openjdk.org/browse/JDK-8293803.
Feel free to modify/comment. :)

**Pull request closed without merging**